--- a/llvm-project/clang/lib/Sema/CheckExprLifetime.cpp
+++ b/llvm-project/clang/lib/Sema/CheckExprLifetime.cpp
@@ -1202,7 +1202,9 @@ static void checkExprLifetimeImpl(Sema &
             << InitEntity->getType()->isReferenceType() << DRE->getDecl()
             << isa<ParmVarDecl>(DRE->getDecl()) << DiagRange;
       } else if (isa<BlockExpr>(L)) {
-        SemaRef.Diag(DiagLoc, diag::err_ret_local_block) << DiagRange;
+        // Don't throw this error during Infer runs
+        // SemaRef.Diag(DiagLoc, diag::err_ret_local_block) << DiagRange;
+        return false;
       } else if (isa<AddrLabelExpr>(L)) {
         // Don't warn when returning a label from a statement expression.
         // Leaving the scope doesn't end its lifetime.
