--- a/llvm-project/clang/lib/Sema/SemaInit.cpp
+++ b/llvm-project/clang/lib/Sema/SemaInit.cpp
@@ -8282,7 +8282,9 @@ void Sema::checkInitializerLifetime(cons
             << Entity.getType()->isReferenceType() << DRE->getDecl()
             << isa<ParmVarDecl>(DRE->getDecl()) << DiagRange;
       } else if (isa<BlockExpr>(L)) {
-        Diag(DiagLoc, diag::err_ret_local_block) << DiagRange;
+        // Don't throw this error during Infer runs
+        // Diag(DiagLoc, diag::err_ret_local_block) << DiagRange;
+        return false;
       } else if (isa<AddrLabelExpr>(L)) {
         // Don't warn when returning a label from a statement expression.
         // Leaving the scope doesn't end its lifetime.
