--- a/b/llvm-project/clang/lib/AST/ItaniumMangle.cpp	2026-03-05 02:25:16.924857372 -0800
+++ b/llvm-project/clang/lib/AST/ItaniumMangle.cpp	2026-03-05 02:29:33.448188783 -0800
@@ -1258,10 +1258,7 @@
 }
 
 void CXXNameMangler::mangleFixedPointLiteral() {
-  DiagnosticsEngine &Diags = Context.getDiags();
-  unsigned DiagID = Diags.getCustomDiagID(
-      DiagnosticsEngine::Error, "cannot mangle fixed point literals yet");
-  Diags.Report(DiagID);
+  Out << "__unhandled_FixedPointLiteral";
 }
 
 void CXXNameMangler::mangleNullPointer(QualType T) {
@@ -4018,11 +4015,7 @@
 }
 
 void CXXNameMangler::mangleNeonVectorType(const DependentVectorType *T) {
-  DiagnosticsEngine &Diags = Context.getDiags();
-  unsigned DiagID = Diags.getCustomDiagID(
-      DiagnosticsEngine::Error,
-      "cannot mangle this dependent neon vector type yet");
-  Diags.Report(T->getAttributeLoc(), DiagID);
+  Out << "__unhandled_NeonVectorType";
 }
 
 static StringRef mangleAArch64VectorBase(const BuiltinType *EltType) {
@@ -4097,11 +4090,7 @@
   Out << TypeName.length() << TypeName;
 }
 void CXXNameMangler::mangleAArch64NeonVectorType(const DependentVectorType *T) {
-  DiagnosticsEngine &Diags = Context.getDiags();
-  unsigned DiagID = Diags.getCustomDiagID(
-      DiagnosticsEngine::Error,
-      "cannot mangle this dependent neon vector type yet");
-  Diags.Report(T->getAttributeLoc(), DiagID);
+  Out << "__unhandled_AArch64NeonVectorType";
 }
 
 // The AArch64 ACLE specifies that fixed-length SVE vector and predicate types
@@ -4195,11 +4184,7 @@
 
 void CXXNameMangler::mangleAArch64FixedSveVectorType(
     const DependentVectorType *T) {
-  DiagnosticsEngine &Diags = Context.getDiags();
-  unsigned DiagID = Diags.getCustomDiagID(
-      DiagnosticsEngine::Error,
-      "cannot mangle this dependent fixed-length SVE vector type yet");
-  Diags.Report(T->getAttributeLoc(), DiagID);
+  Out << "__unhandled_AArch64FixedSveVectorType";
 }
 
 void CXXNameMangler::mangleRISCVFixedRVVVectorType(const VectorType *T) {
@@ -4298,11 +4283,7 @@
 
 void CXXNameMangler::mangleRISCVFixedRVVVectorType(
     const DependentVectorType *T) {
-  DiagnosticsEngine &Diags = Context.getDiags();
-  unsigned DiagID = Diags.getCustomDiagID(
-      DiagnosticsEngine::Error,
-      "cannot mangle this dependent fixed-length RVV vector type yet");
-  Diags.Report(T->getAttributeLoc(), DiagID);
+  Out << "__unhandled_RISCVFixedRVVVectorType";
 }
 
 // GNU extension: vector types
@@ -4845,11 +4826,7 @@
   // instead. See https://lists.isocpp.org/core/2023/04/14118.php
   auto HandleSubstitutionFailure =
       [&](SourceLocation Loc) {
-        DiagnosticsEngine &Diags = Context.getDiags();
-        unsigned DiagID = Diags.getCustomDiagID(
-            DiagnosticsEngine::Error, "cannot mangle this requires-expression "
-                                      "containing a substitution failure");
-        Diags.Report(Loc, DiagID);
+        // Suppress error for Infer
         Out << 'F';
       };
 
@@ -5048,12 +5025,7 @@
   case Expr::BuiltinBitCastExprClass: {
     NotPrimaryExpr();
     if (!NullOut) {
-      // As bad as this diagnostic is, it's better than crashing.
-      DiagnosticsEngine &Diags = Context.getDiags();
-      unsigned DiagID = Diags.getCustomDiagID(DiagnosticsEngine::Error,
-                                       "cannot yet mangle expression type %0");
-      Diags.Report(E->getExprLoc(), DiagID)
-        << E->getStmtClassName() << E->getSourceRange();
+      Out << "__unhandled_BuiltinBitCastExprClass";
       return;
     }
     break;
@@ -5088,12 +5060,7 @@
   // Even gcc-4.5 doesn't mangle this.
   case Expr::BinaryConditionalOperatorClass: {
     NotPrimaryExpr();
-    DiagnosticsEngine &Diags = Context.getDiags();
-    unsigned DiagID =
-      Diags.getCustomDiagID(DiagnosticsEngine::Error,
-                "?: operator with omitted middle operand cannot be mangled");
-    Diags.Report(E->getExprLoc(), DiagID)
-      << E->getStmtClassName() << E->getSourceRange();
+    Out << "__unhandled_BinaryConditionalOperatorClass";
     return;
   }
 
@@ -5463,10 +5430,7 @@
     case UETT_VecStep:
     case UETT_PtrAuthTypeDiscriminator:
     case UETT_DataSizeOf: {
-      DiagnosticsEngine &Diags = Context.getDiags();
-      unsigned DiagID = Diags.getCustomDiagID(
-          DiagnosticsEngine::Error, "cannot yet mangle %0 expression");
-      Diags.Report(E->getExprLoc(), DiagID) << getTraitSpelling(SAE->getKind());
+      Out << "__unhandled_UnaryExprOrTypeTraitExpr";
       return;
     }
     }
@@ -5966,11 +5930,7 @@
         "cannot mangle hlsl temporary value; mangling wrong thing?");
   case Expr::OpenACCAsteriskSizeExprClass: {
     // We shouldn't ever be able to get here, but diagnose anyway.
-    DiagnosticsEngine &Diags = Context.getDiags();
-    unsigned DiagID = Diags.getCustomDiagID(
-        DiagnosticsEngine::Error,
-        "cannot yet mangle OpenACC Asterisk Size expression");
-    Diags.Report(DiagID);
+    Out << "__unhandled_OpenACCAsteriskSizeExpr";
     return;
   }
   }
@@ -6576,9 +6536,7 @@
   // of the data members in the union are unnamed), then there is no way for a
   // program to refer to the anonymous union, and there is therefore no need to
   // mangle its name. However, we should diagnose this anyway.
-  unsigned DiagID = Diags.getCustomDiagID(
-      DiagnosticsEngine::Error, "cannot mangle this unnamed union NTTP yet");
-  Diags.Report(UnionLoc, DiagID);
+  // Suppress error for Infer
 
   return nullptr;
 }
